### ISSUE TITLE
Restore command

### ISSUE THREAD

**Issue by m90** [offen/docker-volume-backup#146]
A container based off this image could also expose a command to restore a volume from a tar archive. This would have to be run manually in the context of the existing Docker setup, providing an archive location (this could probably also be remote).

**Comment by Akruidenberg** [comment]
+1

**Comment by octdanb** [comment]
+1

**Comment by GeneralTao2** [comment]
+1

**Comment by m90** [comment]
Writing down a few thoughts on this before I forget them.

---

### Goals

- Users can restore the contents of a volume from an archive
- The command can be run in a Docker compose setup, using the existing image
- If containers need to be stopped during the procedure, they shall be stopped (just like when backing up)

### Non-Goals

- Fetching archives from remote storages is out of scope for a first version
- Volume mounts and paths will not be figured out automagically, instead users must provide them

### API

This assumes the archive is available on the host that issues the compose command. Is this a problem?

```
docker compose run --rm -v volume_name:/restore backup restore -path /backup/my_app_backup < archive.tar.gz
```

Alternatively the archive could be mounted as well:

```
docker compose run --rm -v ./archive.tar.gz:/tmp/archive.tar.gz -v volume_name:/restore backup restore -path /backup/my_app_backup -archive /tmp/archive.tar.gz
```

### Error handling

How cautious does error handling need to be? Should the command stash the previous contents so it can always roll back to the pre-restoration state?

**Comment by MaxJa4** [comment]
I like the stepped approach: first just concentrate of the copy/restore process using the docker compose command. Add the download etc. later.

> How cautious does error handling need to be? Should the command stash the previous contents so it can always roll back to the pre-restoration state?

As a restore is done in either a testing scenario (non-critical) or when it's actually needed (often critical), any sources of issues should be limited as much as possible imo.

I'm thinking if something like this right now:
- Copy each original object that is being modified by `restore` in any way to a temporary location for potential recovery from failure or also maybe allow aborting and then resuming (would come almost as a "free" feature maybe)
- Maybe also keep a record of modifications (e.g. with status done, ongoing, queued) for the recovery/resume
- Delete the temporary files after a successful restore (all at once at the end or optionally object-by-object to limit storage footprint), otherwise resume or undo the restore

**Comment by MaxJa4** [comment]
Restoring workflow concept (draft, open for discussion):
1. Check if backup-archive is valid / not corrupted
2. Allocate at least as much disk space as the archive is large, maybe even 1.5x or 2x that, to ensure there's (probably) enough disk space
    1. Docker can't find out how much space there is left on the host system, only by trying
    2. Shrink temp file as the backup progresses to keep the space reserved (alternative just do an initial check and then delete it right away)
3. Check if permissions are sufficient to write into the target volume
4. Stop container(s) now or before step 1
5. Copy files
    1. Use atomic copy via creating temp files in the target dir and then `os.rename`. Only atomic in Linux based systems
    2. Option 1: Keep a record of every filesystem change performed by us to be able to recover from an error+retry, abort+retry, crash+retry without having a half restored backup with random files floating around
    3. Option 2: Alternatively purge any found files from a temporary backup to start from a clean slate
    4. For replace-mode: keep a temporary backup of each overwritten/modified file to be able to recover from a failure
    5. For safe-mode: don't copy the files into the target dir yet, use a temp dir in/next to it first
6. Finalizing copying
    1. For replace-mode: Verify checksums, remove temporary backup of original/old files
    2. For safe-mode: move/rename original data away, move/rename restored data to match original data, verify checksums, delete original data
7. Start container(s) again if the were started before

Note: 'replace-mode' and 'safe-mode' are just two approaches. Idk if it makes sense to let the user choose (also bc of vastly different storage requirements; in-place vs. full-size during backup) or to just opt for one these two.
Also, we might do the more complicated stuff like dynamic allocation (2.ii), records for recovery (5.ii) and other things later for a restore v2 including downloading the archive from the specified storage backend or even stream it if its really large (high GB or even TB range).

This is also by no means final or well thought out, more like notes to get a restore strategy built up step by step.

**Comment by m90** [comment]
Some thoughts without having worked through your write up in all detail:

- If we have a proper recovery on failure in place, are all of the pre checks even needed? I would think if a restoration process fails for whatever reason, and the code is able to recover just fine, there's no real need to check anything I would guess. Checks will be incomplete in any case.
- You talk about checksums a lot, where are these coming from? If we don't trust the integrity of backups taken, this should maybe rather be adressed at backup time?
- I like the idea of doing an atomic copy for speed. Is there a Golang library that can do this for us?

I.e. I'd personally maybe focus on a. fast copy / extraction, b. robust recovery.

**Comment by MaxJa4** [comment]
> If we have a proper recovery on failure in place, are all of the pre checks even needed?

Not necessarily, no. It would just safe time. With many gigabytes of data, errors could occur from minutes to hours after starting, as insufficient space or (partially) permissions could lead to an error in the middle or end stage of the recovery. Could also lead to unstable system behavior (storage full).
But to be fair, the user needs to be vary of that to a certain extend. Consider the extended validation as an optional addon for a later point if we feel like it's a benefit to add it.

> You talk about checksums a lot, where are these coming from?

Without any checksums provided in the backup - which would be a nice addition - we could only stream the extracted contents of files through checksum calculation (block-wise for large files which don't fit into memory?) and when writing of that file is done, check if writing was successful and complete (verify with the calculated checksum). But that's arguably quite a lot, complex and too much effort.

> Is there a Golang library that can do this for us?

Atomic copy is basically just doing `os.rename` on Linux based systems. But there is https://github.com/google/renameio which handles that.

**Comment by MaxJa4** [comment]
Updated workflow above, without checksums or crazy amounts of pre-checks. Also visual now, way easier on the eyes - at least for me.

**Comment by buxm** [comment]
I'm hoping to see a restore command implemented soon.
I personally see a backup tool without an easy, quick and tested restore process built in as only half-baked.
Manually fiddling with many tools to fetch the backup from remote, decrypt, extract and restore is prone to error and takes much longer to restore the target system.
